### DISCUSSION TITLE
Mixng "individual" and "json" MQTT Data in one Driver

### DISCUSSION THREAD

**Discussion by jkrasinger** [bellrichm/WeeWX-MQTTSubscribe#176] [General]
I'm using MQTTSubscribe since a few days and it works as expected. Now i like to mix up Data from different Sources (on one MQTT-Instance) which are publishing in a different way. I have data from my "ecowitt" Station via Homeassistant which i receive in "individual" mode, but i like to add some Data from my Netatmo-Station which are presented in "json" mode.

Is it possible to mix-up both data types in the driver?

**Comment by bellrichm** [comment]
Looks like the documentation’ is lagging.  Here is an explanation on how to configure it
https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/113#issuecomment-809655246

**Reply by jkrasinger** [reply]
Thank you for this quick information and excuse me for not reading and searching all the informations, i really did not find it. Especially the part where i can declare the message type individually on each topic. I think it's really missing in the documentation.

Thanks again for very quick help
Jonny

**Comment by peters77** [comment]
Hi,

I use weewx a long time and so pver the years highly customized. I used your subscriber since 2020 I think, I couldn't figure out which version I run!?
I have wrote a litle script that parses some values in a format like and put it into a MQTT server:

`TIME:0,radiation:0.15,cpm:28,cps:2,temp_2m:-1.69,temp_5cm:-2.04,temp_dht22:-0.40,hum_dht22:70.40,rain:0.0`

and it works as a service with this setings in weewx.conf:

```
[MQTTSubscribeService]
    host = 192.168.1.100
    payload_type = keyword

    #    console = false
    #    log = false
    #    [[message_callback]]
    #        type = json
    [[topics]]
        [[[weather]]]    # The topic you are subscribing to
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[temp_2m]]]]    # Name in the MQTT message
                name = extraTemp1    # WeeWX name
            [[[[temp_5cm]]]]
                name = extraTemp2
            [[[[temp_dht22]]]]
                name = tmp_dt_1
            [[[[hum_dht22]]]]
                name = hum_dt_1
            [[[[rain]]]]
                name = hail
            [[[[radiation]]]]
                name = geiger
            [[[[cpm]]]]
                name = cpm
            [[[[cps]]]]
                name = cps

    [[message_callback]]
        type = keyword
        keyword_delimiter = ","    # The delimiter between the name/value pairs
        keyword_separator = :    # The separator between the name and the value

```


Now I would like to add new LoRa sensors which send data to the same MQTT server but it is json formatted payload. I see I can now set a second subscription with a different type (json), like:
....
`{"Bat_status":3.0,"TempC_SHT":17.17,"Ext_sensor":"Temperature Sensor","Hum_SHT":43.6,"TempC_TMP117":16.37,"BatV":3.069},`
....
to pass this into weewx , too!?

Which is the way to update my old stanza and with my given config? Which Version do I need minimum to readout both types in your subscribe service? I see there are some hugh changes like template files setup and the way to install and configure?? 
Any hint to go smoothly on your latest version without breaking my weewx too much!? 
It's a apckage install running weewx 5.1 (debian dpkg).

Thanks in advance!

Regards,

Christian

**Reply by bellrichm** [reply]
It looks like you need at least [2.1.0](https://github.com/bellrichm/WeeWX-MQTTSubscribe/releases/tag/v2.1.0)
The easiest way to see what version you are running is to look in MQTTSubscribe.py and search for `VERSION = `. You should see something like `VERSION = '3.0.0'`.
In theory you should be able to install V3 and it will not update weewx.conf. I will run some tests, but it will take some time. You could try, but make sure you backup weewx.conf (which you would do on any install anyway….)

If you want to try some of the V3 functionality, you could create a test stanza and use that to check that the configuration is correct. Or you could just edit weewx.conf in place (after making a backup), restart WeeWX and see what happens…

The updates in V3 are to (hopefully) make it easier for new users to configure MQTTSubscribe and reduce the support load. But it is all optional.

Take a look at this, https://github.com/bellrichm/WeeWX-MQTTSubscribe/blob/ce84e6043779f24f5e970c1b949f5050def6ad17/mqttsubscribe.example.conf#L80
Notice how the hierarchy is `[[topics]] [[[TOPIC-NAME]]] [[[[message]]]]`. This is the format you need to use to support different ‘topic types’

**Reply by peters77** [reply]
Thanks for the quick reply.
I have 1.5.4-rc01 installed and running, so out of date a lot!  👎 
Should I follow this guide: 
'Installing and Updating Version 2.X and Earlier' and install 2.1 in top?
Will do a copy of weewx.conf before in any case.

Or can I just copy and replace some files in bin/usr!?

**Reply by bellrichm** [reply]
Well….. Yes, you do have a lot of options. 
While, probably not recommended, you could just grab v3 and replace MQTTSubscribe.py (after a backup of it). A quick glance of your weewx.conf, it looks like it should be compatible. I would then restart MQTTSubscribe and getting it working. I would then update it remove the `message_callback` section and add the `message` section under the current topic. Once that is working, I would add the new topic. 
I think one drawback of this is, that the WeeWX install (formerly wee_extension, currently weectl) keeps track of the version of the extension…

It comes down to how comfortable you are ‘going off the beaten path’.

Rich

**Reply by bellrichm** [reply]
You might want to review this, https://github.com/bellrichm/WeeWX-MQTTSubscribe/wiki/Options-removed-in-2.0.0
Again, I **think** you should be fine with the current weewx.conf with 2.x or 3.x…

**Reply by bellrichm** [reply]
Something like this should work with v2 or v3
```
[MQTTSubscribeService]
    host = 192.168.1.100
    payload_type = keyword

    #    console = false
    #    log = false
    #    [[message_callback]]
    #        type = json
    [[topics]]
        [[[weather]]]    # The topic you are subscribing to
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[message]]]]
                type = keyword
                keyword_delimiter = ","    # The delimiter between the name/value pairs
                keyword_separator = :    # The separator between the name and the value

            [[[[temp_2m]]]]    # Name in the MQTT message
                name = extraTemp1    # WeeWX name
            [[[[temp_5cm]]]]
                name = extraTemp2
            [[[[temp_dht22]]]]
                name = tmp_dt_1
            [[[[hum_dht22]]]]
                name = hum_dt_1
            [[[[rain]]]]
                name = hail
            [[[[radiation]]]]
                name = geiger
            [[[[cpm]]]]
                name = cpm
            [[[[cps]]]]
                name = cps

**Reply by peters77** [reply]
Ok....will do an additional full backup of weewx and try to start the adventure....leaving the comfort zone...! 👍  :-D

**Reply by bellrichm** [reply]
It probably sounds worse/harder than it is….

**Reply by peters77** [reply]
Rich,

ok....replaced my old MQTTSubscribe.py with the new one, it's working with the latest V3 version. Pasted your provided adapted section into weewx.conf. All working! 👍 
Got one error on startup, but values are fetched and updated:
`
ERROR: Unknown option: MQTTSubscribe-payload_type`

Sould I now first do a full udpate/upgrade with weectl or am I fine for now!? :-D 
Next step would be adding the more complicated second topic...

Regards,

Christian

**Reply by peters77** [reply]
Ok...maybe 
```

[MQTTSubscribeService]
    host = 192.168.1.100
    payload_type = keyword
```
the     `payload_type = keyword` line should be removed as it will now be set under the topics section..!?
```

 [[topics]]
        [[[weather]]]    # The topic you are subscribing to
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[message]]]]
                type = keyword
```

..? ....ok...removed, error is gone... 👍

**Reply by bellrichm** [reply]
Yup, `payload_type` was an old opyion.
I’d get the second topic working

**Reply by peters77** [reply]
Ok, will try...but difficult to debug as the sensor only emitts data every 20min! Never worked with parsing json. And the topic seems quite complicated. Am I right that if I only get a debug message in weewx log if there is an incoming topic and the subscription works...!?

**Reply by bellrichm** [reply]
Yes and No. It does mean that MQTTSubscribe worked. But to be sure WeeWX has what you want, you really need to check the loop packets/archive records and the database.

Using the attatched files, I ran MQTTSubscribe in [parser mode](https://github.com/bellrichm/WeeWX-MQTTSubscribe/wiki/Parser-mode).
The exact invocation I used was: 
```python3 $USER_ROOT/MQTTSubscribe.py parse $RUN_MODE --top-level --conf mqttsubscribe.conf --message-file topic.json --topic loran```

This gave the following the output:
```{'loran_battery_status': 3.0, 'loran_temp1': 17.17, 'loran_sensor': 'Temperature Sensor', 'loran_humidity': 43.6, 'loran_temp2': 16.37, 'loran_battery_voltage': 3.069, 'dateTime': 1737463843.6788425, 'usUnits': 17}```

You should just need to configure the appropriate topic and WeeWX names.
Hope this helps. rich

[mqttsubscribe.conf.txt](https://github.com/user-attachments/files/18491182/mqttsubscribe.conf.txt)
[topic.json](https://github.com/user-attachments/files/18491193/topic.json)

**Reply by peters77** [reply]
Ok, I made a mosq.conf, here is my config:

```
[MQTTSubscribeService]
    host = 192.168.1.100

    #    console = false
    #    log = false
    #    [[message_callback]]
    #        type = json
    [[topics]]
        ignore_start_time = True
        [[[weather]]]    # The topic you are subscribing to
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[message]]]]
                type = keyword
                keyword_delimiter = ","    # The delimiter between the name/value pairs
                keyword_separator = :    # The separator between the name and the value

            [[[[temp_2m]]]]    # Name in the MQTT message
                name = extraTemp1    # WeeWX name
            [[[[temp_5cm]]]]
                name = extraTemp2
            [[[[temp_dht22]]]]
                name = tmp_dt_1
            [[[[hum_dht22]]]]
                name = hum_dt_1
            [[[[rain]]]]
                name = hail
            [[[[radiation]]]]
                name = geiger
            [[[[cpm]]]]
                name = cpm
            [[[[cps]]]]
                name = cps

        [[[application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f2c1799a2f/event/up]]]    # The topic you are subscribing to
           ignore = true
            [[[[message]]]]
                type = json
#                keyword_delimiter = ","    # The delimiter between the name/value pairs
#                keyword_separator = :    # The separator between the name and the value

            [[[[object_TempC_TMP117]]]]    # Name in the MQTT message
                ignore = false # opt in for this field
                name = extraTemp4    # WeeWX name
            [[[[object_TempC_SHT]]]]
                ignore = false # opt in for this field
                name = extraTemp5
            [[[[object_Hum_SHT]]]]
                ignore = false # opt in for this field
                name = extraHumid4
            [[[[object_BatV]]]]
                ignore = false # opt in for this field
                name = batteryStatus4
```

after some fiddeling and parsing the quite complicated payload with your parser mode shows:

```
(Service) Skipping rxInfo because data is an array and has no configured subfields. Array=[{'gatewayId': '4b364b5156414b32', 'uplinkId': 64513, 'gwTime': '2025-01-21T19:28:43.155867+00:00', 'nsTime': '2025-01-21T19:28:44.907126188+00:00', 'rssi': -41, 'snr': 9.75, 'channel': 7, 'location': {}, 'context': 'UuqRzA==', 'metadata': {'region_common_name': 'EU868', 'region_config_id': 'eu868'}, 'crcStatus': 'CRC_OK'}]
{'extraTemp5': 17.46, 'extraHumid4': 42.3, 'batteryStatus4': 3.064, 'extraTemp4': 17.38, 'dateTime': 1737490945.7933917, 'usUnits': 1}
```

So this config seems to parse the two temp and the hum values, but the rxInfo should be somehow masked out!? 
 I included for testing this in my index.html.tmpl:

```
                      <tr>
                        <td>Temperatur 5cm Lora</td>
                        <td style="font-size:120%;color:#343a40">$current.extraTemp4</td>
                      </tr>

```
but get always

**Temperatur 5cm Lora 	N/A**

My cycle is 1min. 
So on the way but `extraTemp4` is not updating at the moment by mqtt. 
I added `extraTemp4` new to the database but tested if I can write values in it and it is working, so not a problem from the database I think. Is the value lost after the next 1min cycle if there is only data read every 20min or should the last value show up till updating by the next mqqt input cycle? 

But won't bother you to much with all this stuff! 

Regards,

Christian 

P.S.: I just got this in the weewx.log:

```
2025-01-21T22:08:43.604843+01:00 weewx weewxd[831538]: DEBUG user.MQTTSubscribe: (Service) TopicManager data-> incoming application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f7c1899a2f/event/up: 'batteryStatus4': '3.065', 'extraHumid4': '42.6', 'extraTemp4': '17.24', 'extraTemp5': '17.31'
2025-01-21T22:08:43.863518+01:00 weewx weewxd[831538]: DEBUG user.MQTTSubscribe: (Service) TopicManager data-> outgoing application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f7c1899a2f/event/up: 'batteryStatus4': '3.065', 'dateTime': '1737493723.6046684', 'extraHumid4': '42.6', 'extraTemp4': '17.24', 'extraTemp5': '17.31', 'usUnits': '1'
2025-01-21T22:08:43.864384+01:00 weewx weewxd[831538]: DEBUG user.MQTTSubscribe: (Service) TopicManager data-> outgoing accumulated application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f7c1899a2f/event/up: 'batteryStatus4': '3.065', 'dateTime': '1737493724.0', 'extraHumid4': '42.6', 'extraTemp4': '17.24', 'extraTemp5': '17.31', 'usUnits': '1'
```

In that one cycle where data is coming in - the updated HP shows now:

Temperatur 5cm Lora 	-8,2°C

Hm, the value is wrong, redout is 17.24 Grad Celsius ??

**Reply by bellrichm** [reply]
Could you attach/embed your MQTT payload? I'd like to take a look at the parsing error you are getting. (Probably good for completeness to attach/embed your MQTTSubscribe configuration too.)
re: temp incorrect
Your new topic is missing `unit_system = METRICWX`

I'll circle back to the timing of receiving MQTT messages.

**Reply by peters77** [reply]
...yes...just added `unit_system = METRICWX` and now the value is displayed right (for that 1min circle). 👍 

That's my payload from ChirpStack V4 with MQTT, Mikrotik Gateway and Dragino LHT65N

`{"deduplicationId":"2c90346b-ffa2-4236-a85a-99800964aafd","time":"2025-01-21T21:48:40.985912+00:00","deviceInfo":{"tenantId":"52f14gd4-c8f1-4fed-8f87-4025e1d49242","tenantName":"ChirpStack","applicationId":"d11a30e3-f87a-4866-949d-e07ab9aad18e","applicationName":"temp-hum-app","deviceProfileId":"9f4e5031-3788-474a-b996-23471906dee9","deviceProfileName":"Temperatur+Feuchte","deviceName":"LHT65N","devEui":"a84021f7c1898a2f","deviceClassEnabled":"CLASS_A","tags":{}},"devAddr":"012d10aa","adr":true,"dr":5,"fCnt":177,"fPort":2,"confirmed":false,"data":"y/cGuQGpAgayf/8=","object":{"Bat_status":3.0,"TempC_SHT":17.21,"Hum_SHT":42.5,"TempC_TMP117":17.14,"Ext_sensor":"Temperature Sensor","BatV":3.063},"rxInfo":[{"gatewayId":"4a365b5192314b12","uplinkId":26369,"gwTime":"2025-01-21T21:48:40.985912+00:00","nsTime":"2025-01-21T21:48:42.783047969+00:00","rssi":-43,"snr":9.75,"channel":7,"location":{},"context":"R3hGLA==","metadata":{"region_common_name":"EU868","region_config_id":"eu868"},"crcStatus":"CRC_OK"}],"txInfo":{"frequency":867900000,"modulation":{"lora":{"bandwidth":125000,"spreadingFactor":7,"codeRate":"CR_4_5"}}}}`

That's my current MQTTSubscribe configuration:

```
[MQTTSubscribeService]
    host = 192.168.1.195

    #    console = false
    #    log = false
    #    [[message_callback]]
    #        type = json
    [[topics]]
        ignore_start_time = True
        [[[weather]]]    # The topic you are subscribing to
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[message]]]]
                type = keyword
                keyword_delimiter = ","    # The delimiter between the name/value pairs
                keyword_separator = :    # The separator between the name and the value

            [[[[temp_2m]]]]    # Name in the MQTT message
                name = extraTemp1    # WeeWX name
            [[[[temp_5cm]]]]
                name = extraTemp2
            [[[[temp_dht22]]]]
                name = tmp_dt_1
            [[[[hum_dht22]]]]
                name = hum_dt_1
            [[[[rain]]]]
                name = hail
            [[[[radiation]]]]
                name = geiger
            [[[[cpm]]]]
                name = cpm
            [[[[cps]]]]
                name = cps

        [[[application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f7c1899a2f/event/up]]]    # The topic you are subscribing to
            ignore = true
            unit_system = METRICWX    # Not sure weather you want METRICWX or METRIC. See http://weewx.com/docs/customizing.htm#units
            [[[[message]]]]
                type = json

            [[[[object_TempC_TMP117]]]]    # Name in the MQTT message
                ignore = false # opt in for this field 
                name = extraTemp4    # WeeWX name
            [[[[object_TempC_SHT]]]]
                ignore = false # opt in for this field 
                name = extraTemp5
            [[[[object_Hum_SHT]]]]
                ignore = false # opt in for this field 
                name = extraHumid4
            [[[[object_BatV]]]]
                ignore = false # opt in for this field 
                name = batteryStatus4
```

**Reply by bellrichm** [reply]
The error message is related to this [issue](https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/219). To hide/suppress it you can use this hack. Add the following as a field under the `application/d11a28e3-f97a-4866-949d-e07ab9aad18e/device/a84041f7c1899a2f/event/up` topic.

```
           [[[[rxInfo]]]]
                [[[[[subfields]]]]]
                    [[[[[[gatewayId]]]]]]
                        ignore = true
```

**Reply by peters77** [reply]
Thanks, added it to my config. 
Hm, I tried to plot the data, but as the data arrives every 20min the plot shows only dots, no line netween the dots, but this seems to be related to this:

https://groups.google.com/g/weewx-user/c/ML6HPMmCXbs

I think I have to live with cross type and gaps in the plot
.
<img width="321" alt="Bildschirmfoto 2025-01-22 um 11 26 16" src="https://github.com/user-attachments/assets/a27d04db-4b20-447c-ad7f-b190e3a1e4b8" />


Event tried to alter my code to readout from `$current` to `$latest`:
```
                     <tr>
                        <td>Temperatur 5cm Lora</td>
##                        <td style="font-size:120%;color:#343a40">$current.extraTemp4</td>
                        <td style="font-size:120%;color:#343a40">$latest.extraTemp4</td>
                      </tr>
```
But still only a value visible in the 1min loop I got data in, the other 19min N/A. 
As Lora and the bigger data cycles become more common I hope there will be some take care of in WeeWx in the future. 
<img width="344" alt="Bildschirmfoto 2025-01-22 um 11 09 30" src="https://github.com/user-attachments/assets/afd9df25-bcff-430b-a41c-53810fa4258b" />

<img width="349" alt="Bildschirmfoto 2025-01-22 um 11 10 38" src="https://github.com/user-attachments/assets/5ca21255-0988-4591-915f-083b938e3eba" />
 
I could try to set the uplink transmit period to 1 min to get in each cycle a value, but it will take much more battery. 
Would be nice to have some function to insert the latest readout value into the loop/archive paket if there is no new reaout to get a constant value!? 
But just an idea.

**Reply by bellrichm** [reply]
I had attempted to implement caching archive records values. I learned a lot about the internals of WeeWX while doing so. The implementation ended up having the limitation that MQTTSubscribeService needs to bind to the archive record. The one person who had interest in the function could live with this…
I opened [issue 178](https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/178), but nobody has voiced a need/desire, so I haven’t actively started to work on it. It might be time to look at it again…

**Reply by peters77** [reply]
Caching sounds really good. +1 for implementing chaching like in https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/178 porposed and descibed. Would be nice to have a smoothed graph and values in each archive record and each updated website in weewx. 
If you find the time it would be a great improvement.

**Reply by bellrichm** [reply]
I’m looking into this. The hardest part will be testing it….
Let’s move all future discussion on it to [issue 178](https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/178)
